### ISSUE TITLE
nvim_buf_set_lines move cursor when there are multiple window

### ISSUE THREAD

**Issue by ipod825** [neovim/neovim#7756]
- `nvim --version`: 0.2.3-dev
- Operating system/version: android
- Terminal name/version: termux
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`
With test.txt as
```
0
1
```
and the file test.vim as
```vim
let buf = nvim_get_current_buf()
call nvim_command('vsplit')
call nvim_command('wincmd l')
call nvim_feedkeys('j', 'n', v:false)
call nvim_buf_set_lines(buf, 1, 2, v:false, ['2'])
```
Start nvim with
```
nvim -u NORC test.txt
```
```vim
:source test.vim
```
That would lead to two window displaying test.txt with modified content:
```
0
2
```

### Expected behaviour
The cursor of the left window is at line 0 and the cursor of the right window is at line 1. 

### Actual behaviour
However, if we do the same steps but type the commands in `test.vim` manually instead of sourcing it, the cursor on the right windo will be at line 0, i.e., calling `nvim_buf_set_lines` move the cursor of the second window.

**Comment by ethanboswell00** [comment]
Hey, I am interested in working on this issue as a first timer. May I contribute?

**Comment by justinmk** [comment]
@ethanboswell00 Absolutely. Read [CONTRIBUTING.md]( https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md) to get started. 

See [run the locally](https://github.com/neovim/neovim/blob/master/test/README.md#running-tests) for help with writing & running tests. 

tl;dr: running a single test file works like this:

    TEST_FILE=test/functional/foo_spec.lua make functionaltest

**Comment by phodge** [comment]
@ethanboswell00: I'm familiar with these parts of the code as well so feel free to ping me here or on the neovim gitter if you have questions.

**Comment by phodge** [comment]
See also: https://github.com/neovim/neovim/issues/8659 - it might be worthwhile fixing that at the same time.

**Comment by ethanboswell00** [comment]
I understand the issue and I've studied the code for the last couple of days, but am still stumped as to what code to change. The issue is with the cursor after the nvim_buf_set_lines function is called, but I can't tell what connection the cursor has to events being called from the prompt from within the code. I've made numerous cat | grep calls in the repository just to track something down, but with no luck. Am I missing something here? Which course of action would I best take to alter event handling with the cursor in Neovim?

**Comment by phodge** [comment]
I've just been playing with some test files that have more than two lines. I have observed that:

- when you call `nvim_buf_set_lines()`, any other windows with their cursor on that line ahve their cursor moved up
- there is special code just for the current window such that its cursor will be put back in case it gets moved by `nvim_buf_set_lines()`
- the call to `switch_to_win_for_buf()` doesn't check to see if you're already inside a window for the buffer, so it sometimes puts you in another window while the change is made. (I think it always puts you in the lowest-numbered window of the lowest-numbered tab containing the buffer.) Therefore sometimes your window will get the cursor-preserving behaviour, and sometimes it gets the cursor-moved-up behaviour, depending on what other windows you have open for that buffer

**Comment by phodge** [comment]
@ethanboswell00: now that I'm reading the comments again, I _think_ the fix for this is just to modify `switch_to_win_for_buf()` so that it won't switch windows if the current window already points to the requested buffer. It's probably a 2-line fix - are you still interested in submitting a PR for it?

**Comment by ethanboswell00** [comment]
Wow, thank you for taking another look through the code and catching all the aspects I missed. I haven't completed this yet since I was away from my computer earlier this week, but I will still make the edit and submit a PR.

**Comment by shaform** [comment]
> I think the fix for this is just to modify switch_to_win_for_buf() so that it won't switch windows if the current window already points to the requested buffer. 

But this only fix the cursor of the current window, other windows with the same buffer will still have their cursors moved.

**Comment by justinmk** [comment]
Fixed by #9674.

**Comment by ipod825** [comment]
This is not fixed entirely. 
nvim --version: NVIM v0.4.0-819-g66d127957

With test.txt as
```
0
1
2
```
and run the the following commands after opening after `nvim -U NORC test.txt'
```vim
let buf = nvim_get_current_buf()
call nvim_feedkeys('j', 'n', v:false)
call nvim_command('vsplit')
call nvim_command('wincmd l')
call nvim_feedkeys('j', 'n', v:false)
call nvim_buf_set_lines(buf, 1, 2, v:false, ['2'])
```
Ideally, the cursor for the left window should stay at the second line. But it is set to the first line.

**Comment by bfredl** [comment]
@ipod825 The semantics of`nvim_buf_set_lines` is that it removes the entire line range and then adds new lines that have no relation to the old ones, it is not the "same" line anymore. For the behavior you want what is needed is some new API function that can make changes within lines, for instance replace the text between two arbitrary marks. This would preserve cursor positions and marks in untouched parts of the line, including just after a NL char if it is not replaced.

**Comment by ipod825** [comment]
@bfredl 
Thanks for the explanation. Currently setting a single line of the buffer via python [api](https://github.com/neovim/pynvim/blob/4dd9962c47bc9d48dc01d7cdf99e3c9819e97548/pynvim/api/buffer.py#L61):
```python
nvim.current.buffer[0] = 'new text'
```
calls `nvim_buf_set_lines`. So do you have any suggestion for setting a single line without moving the cursor?

**Comment by ipod825** [comment]
Any thoughts? Or is it just a missing in the current implementation?

**Comment by justinmk** [comment]
see https://github.com/neovim/neovim/issues/7756#issuecomment-496834966 , would need a new API function.

**Comment by ipod825** [comment]
O.K. I'll open a new issue for it. Thanks.
